core/serialization/download_xlsx_serializer.py

```diff
@@ -0,0 +1,660 @@
+    funding = get_funding(project_ids, places_metadata, organisation_metadata)
+    funding_comments = get_funding_comments(project_ids, places_metadata, organisation_metadata)
+    private_investments = get_private_investments(project_ids, places_metadata, organisation_metadata)
+    outputs_ref = [OutputDimSerializer(output_dim).to_dict() for output_dim in ents.OutputDim.query.all()]
+    output_data = get_outputs(project_ids, places_metadata, organisation_metadata)
+    outcome_ref = [OutcomeDimSerializer(outcome_dim).to_dict() for outcome_dim in ents.OutcomeDim.query.all()]
+    project_outcomes = [
+        OutcomeDataSerializer(
+            outcome_data,
+            places_metadata[
+                outcome_data.programme.programme_id if outcome_data.programme else outcome_data.project.project_id
+            ],
+            organisation_metadata[
+                outcome_data.programme.programme_id if outcome_data.programme else outcome_data.project.project_id
+            ],
+        ).to_dict()
+        for outcome_data in project_outcomes
+    ]
+
+    risks = get_risks(
+        programme_ids, project_ids, places_metadata, organisation_metadata
+    )  # risks are combination of programme and project risks
+    outcomes = [*programme_outcomes, *project_outcomes]  # outcomes are combination of programme and project outcomes
+
+    # Provides sheet order for when outputted in Excel format.
+    download_data = {
+        "PlaceDetails": place_details,
+        "ProjectDetails": project_details,
+        "OrganisationRef": organisation_refs,
+        "ProgrammeRef": programme_refs,
+        "ProgrammeProgress": programme_progresses,
+        "ProjectProgress": project_progresses,
+        "FundingQuestions": funding_questions,
+        "Funding": funding,
+        "FundingComments": funding_comments,
+        "PrivateInvestments": private_investments,
+        "OutputsRef": outputs_ref,
+        "OutputData": output_data,
+        "OutcomeRef": outcome_ref,
+        "OutcomeData": outcomes,
+        "RiskRegister": risks,
+    }
+    return download_data
+
+
+def org_metadata_programme_id(programme_ids):
+    organisations = {}
+
+    programmes = (
+        ents.Programme.query.filter(ents.Programme.id.in_(programme_ids))
+        .options(
+            joinedload(ents.Programme.organisation),  # preload org
+            joinedload(ents.Programme.projects),  # preload child projects
+        )
+        .all()
+    )
+
+    for programme in programmes:
+        # add the programme -> org mapping
+        organisations[programme.programme_id] = programme.organisation.organisation_name
+
+        # add the project -> org mapping for each child project of programme
+        for project in programme.projects:
+            organisations[project.project_id] = programme.organisation.organisation_name
+
+    del programmes  # don't rely on the garbage collector, just delete now to preserve memory
```

Comment:
was thinking to just use this as a precaution instead of relying on the garbage collector but thinking about it more I agree, I don't think it is necessary / probably wont make any difference. It's not memory that is the issue and garbage collector should just clean it up as soon as the function ends anyway...